### DISCUSSION TITLE
perf event array: failed to create perf ring for CPU XYZ: can't create perf event: invalid argument

### DISCUSSION THREAD

**Discussion by lmb** [cilium/ebpf#1029] [Q&A] [answered]
Some users [run into problems](https://github.com/cilium/ebpf/issues/968) when trying to open a perf reader for a perf event array.

Given a map definition like this:

```C
#define EVENT_MAX_ENTRIES       512

struct bpf_elf_map __section_maps event = {
    .type = BPF_MAP_TYPE_PERF_EVENT_ARRAY,
    .size_key = sizeof(int),
    .size_value = sizeof(__u32),
    .max_elem = EVENT_MAX_ENTRIES,
    .pinning = PIN_GLOBAL_NS,
};
```

Given code like this:

```go
m, err := ebpf.LoadPinnedMap("/path/to/map", nil)
if err != nil {
	fmt.Printf("err: %v\n", err)
	return
}

reader, err := perf.NewReader(m, 8192)
if err != nil {
	fmt.Printf("err: %v\n", err)
	return
}
```

Will generate an error like `failed to create perf ring for CPU 32: can't create perf event: invalid argument`.

**Comment by lmb** [comment, marked as answer]
The problem is that the map size is larger than the total number of possible CPUs in the system. In the example the system probably has 16 cores / 32 hyperthreads numbered from 0-31, hence the error when trying to create a perf ring for CPU 32.

The fix is to not specify `max_elems` for perf event arrays, which makes the loader determine the correct size for the current system. This is supported both by cilium/ebpf and [newer libbpf / iproute2 versions](https://github.com/libbpf/libbpf/blob/b5e9722ec268702b2ec62863222277148e4fe656/src/libbpf.c#L5184).

**Comment by shach33** [comment]
I am not specifying the max entries at all.  My definition looks like 

```
struct {
__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
__uint(size_key, sizeof(__u32));
__uint(size_value, sizeof(__u32));
 
} payload SEC(".maps");
```

```
$ sudo bpftool map show name "payload"

563: perf_event_array  name payload  flags 0x0
        key 4B  value 4B  max_entries 64  memlock 4096B
        pids process1-ebpf-(1768522)
```

My machine has 80 CPUs, and when I run `NewReaderWithOptions  (payload, os.Getpagesize())`  the error  I get is 

`failed to create perf ring for CPU 80: can't create perf event: invalid argument `

The libbpf version is  1:0.5.0-1ubuntu22.04.1 .

**Comment by shach33** [comment]
I was able to fix it. The Maxentries were being set due to a variable assignment error. Please close this.
